### PR TITLE
FEATURE: Allow dynamic newReferenceNodePath

### PR DESCRIPTION
**Pull request by Sebobo**
With this change the `newReferenceNodePath`
setting will be parsed with eel and is therefore
adjustable via a flowquery.

### CODE REVIEW

**Comment by Sebobo** [comment]
Ok now it doesn't fetch it if the path starts with a "/".

I feel it's somehow a bit messy with the eel parsing, but I'm not sure how to improve it right now.
And I tried using actual eel in the Settings including the `${}` but for some reason it get's parsed immediately and creates an error. Then it would be easier to detect.

File: Resources/Private/FlatNav/src/FlatNav.js [hunk 1]
```diff
@@ -139,13 +139,15 @@ class FlatNav extends Component {
...
+            if (this.props.preset.newReferenceNodePath.indexOf('/') !== 0) {
+                this.props.fetchNewReferenceNodePath();
+            }
         }
     }
 
     createNode = () => {
         const context = this.props.siteNodeContextPath.split('@')[1];
-        const contextPath = this.props.newReferenceNodePath + '@' + context;
+        const contextPath = (this.props.newReferenceNodePath || this.props.preset.newReferenceNodePath) + '@' + context;
```
Comment (dimaip) [inline Resources/Private/FlatNav/src/FlatNav.js]: What happens when you try to create a node before the `fetchNewReferenceNodePath` has finished loading?
I'd block the "+" button until the request is complete.

**Comment by Sebobo** [comment]
Plus button is now hidden while the reference path loads.

Is there any nice way to not define the whole preset state fully all the time?

**Comment by dimaip** [comment]
> Is there any nice way to not define the whole preset state fully all the time?

Sure. E.g. use immutable data or deep merge. `setState` does shallow merge.
E.g. see https://stackoverflow.com/questions/24898012/react-js-setstate-overwriting-not-merging

The way I implemented it is obviously sloppy, was just a quick code jam.

**Comment by dimaip** [comment]
> I feel it's somehow a bit messy with the eel parsing, but I'm not sure how to improve it right now.
And I tried using actual eel in the Settings including the ${} but for some reason it get's parsed immediately and creates an error. Then it would be easier to detect.

Yes, exactly, that's why I used Eel expressions without ${} originally.

File: Resources/Private/FlatNav/src/FlatNav.js [hunk 2]
```diff
@@ -55,33 +59,72 @@ const makeFlatNavContainer = OriginalPageTree => {
...
+        makeGetNewReferenceNodePath = preset => () => {
+            this.setState({
+                [preset]: {
+                    page: this.state[preset].page,
+                    isLoading: this.state[preset].isLoading,
+                    isLoadingReferenceNodePath: true,
+                    nodes: this.state[preset].nodes,
+                    moreNodesAvailable: this.state[preset].moreNodesAvailable,
+                    newReferenceNodePath: this.state[preset].newReferenceNodePath
+                }
```
Comment (dimaip) [inline Resources/Private/FlatNav/src/FlatNav.js]: perhaps the easiest fix not to overwrite everything is to use spread:

```
this.setState({
  [preset]: {
    ...this.state[preset],
    isLoadingReferenceNodePath: true
  }
})
```

**Review by dimaip — CHANGES_REQUESTED** [review]

**Comment by Sebobo** [comment]
@dimaip thanks that made it much better and safer for future enhancements

**Comment by dimaip** [comment]
Thanks!!!

**Pull request merged**